rust/src/delta.rs

```diff
@@ -1188,99 +1177,22 @@ impl DeltaTable {
-        Ok((path_name.starts_with('.') || path_name.starts_with('_'))
-            && !path_name.starts_with("_delta_index")
-            && !path_name.starts_with("_change_data")
-            && !self
-                .state
-                .current_metadata()
-                .ok_or(DeltaTableError::NoMetadata)?
-                .partition_columns
-                .iter()
-                .any(|partition_column| path_name.starts_with(partition_column)))
-    }
-
-    /// Run the Vacuum command on the Delta Table: delete files no longer referenced by a Delta table and are older than the retention threshold.
-    /// We do not recommend that you set a retention interval shorter than 7 days, because old snapshots
-    /// and uncommitted files can still be in use by concurrent readers or writers to the table.
-    /// If vacuum cleans up active files, concurrent readers can fail or, worse, tables can be
-    /// corrupted when vacuum deletes files that have not yet been committed.
-    /// If `retention_hours` is not set then the `configuration.deletedFileRetentionDuration` of
-    /// delta table is used or if that's missing too, then the default value of 7 days otherwise.
+    /// Vacuum the delta table seee [`Vacuum`] for more info
     pub async fn vacuum(
-        &self,
+        &mut self,
         retention_hours: Option<u64>,
         dry_run: bool,
-        enforce_retention_duration: Option<bool>,
-    ) -> Result<Vec<String>, DeltaTableError> {
-        let expired_tombstones =
-            self.get_stale_files(retention_hours, enforce_retention_duration)?;
-        let valid_files = self.get_file_set();
-
-        let mut files_to_delete = vec![];
-        let mut all_files = self.storage.list_objs(&self.table_uri).await?;
-
-        while let Some(obj_meta) = all_files.next().await {
-            let obj_meta = obj_meta?;
-            let rel_path = extract_rel_path(&self.table_uri, &obj_meta.path)?;
-
-            if valid_files.contains(rel_path) // file is still being tracked in table
-                || !expired_tombstones.contains(rel_path) // file is not an expired tombstone
-                || self.is_hidden_directory(rel_path)?
-            {
-                continue;
-            }
-
-            files_to_delete.push(obj_meta.path);
+        enforce_retention_duration: bool,
+    ) -> Result<Vec<String>, VacuumError> {
+        let mut plan = Vacuum::default()
+            .dry_run(dry_run)
+            .enforce_retention_duration(enforce_retention_duration);
+        if let Some(hours) = retention_hours {
+            plan = plan.with_retention_period(Duration::hours(hours as i64));
         }
 
-        if dry_run {
-            return Ok(files_to_delete);
-        }
-        match self.storage.delete_objs(&files_to_delete).await {
-            Ok(_) => Ok(files_to_delete),
-            Err(err) => Err(DeltaTableError::StorageError { source: err }),
-        }
+        let res = plan.execute(self).await?;
```

Comment:
One of my hopes in proposing #661 was that we could eventually return the builder and arrive at an API like.

```rs
let dt = open_table(...);
let result = dt.vacuum().enforce_retention_duration(true).into_future().await;
```

once the [`IntoFuture`](https://doc.rust-lang.org/std/future/trait.IntoFuture.html) trait lands, this would simplify to.

```rs
let dt = open_table(...);
let result = dt.vacuum().enforce_retention_duration(true).await;
```

This would also mean we don't have to have vacuum itself be async anymore, but requires that the `VacuumPlan` has everything it needs to execute. I left some thoughts in to top level comment.

Reply:
Yeah if `DeltaLog` was implemented the `into_future()` change would be trivial.  I'll play on a nightly branch and see how hard it would be without the log.